streaming/src/event_server.h

```diff
@@ -0,0 +1,216 @@
+#ifndef RAY_STREAMING_EVENT_SERVER_H
+#define RAY_STREAMING_EVENT_SERVER_H
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <thread>
+#include <unordered_map>
+
+#include "channel.h"
+#include "ring_buffer.h"
+#include "util/streaming_util.h"
+
+namespace ray {
+namespace streaming {
+
+enum class EventType : uint8_t {
+  UserEvent = 0,
+  FlowEvent = 1,
+  EmptyEvent = 2,
+  FullChannel = 3,
+  Reload = 4,
+};
+
+struct EnumTypeHash {
+  template <typename T>
+  std::size_t operator()(const T &t) const {
+    return static_cast<std::size_t>(t);
+  }
+};
+
+struct Event {
+  ProducerChannelInfo *channel_info;
+  EventType type;
+  bool urgent;
+};
+
+/// Data writer utilizes what's called an event-driven programming model instead
+/// loop forward polling driven that's employed in first simple version.
+/// Event driven model includes two important components: event server and event
+/// queue. In the process of data transmission, the inputer will first define
+/// the processing method of corresponding events. However, by triggering
+/// different events in actual operation, these events will be put into the event
+/// queue, and finally the event server will schedule the previously registered
+/// processing functions ordered by its priority.
+
+/// The event queue inherits from the universal ring queue. It mainly divides
+/// event into two different levels: normal event and urgent event, and the
+/// total size of the queue is the sum of them.
+class EventQueue : public AbstractRingBuffer<Event> {
```

Comment:
`EventQueue` seems not to be a `AbstractRingBuffer`? It's just a blocking queue, not a ring buffer. Is that right?

Reply:
> `EventQueue` seems not to be a `AbstractRingBuffer`? It's just a blocking queue, not a ring buffer. Is that right?

It's fixed size queue base ringbuffer.

Reply:
Can you elaborate more about `ring`? I don's see any `ring` here. It's a fixed size blocking queue. I don't see any wrap around from tail to head.

Reply:
Fixed